Tools/peg_generator/pegen/c_generator.py

```diff
@@ -302,44 +287,62 @@ def visit_Opt(self, node: Opt) -> FunctionCall:
             comment=f"{node}",
         )
 
-    def visit_Repeat0(self, node: Repeat0) -> FunctionCall:
-        if node in self.cache:
-            return self.cache[node]
-        name = self.gen.artificial_rule_from_repeat(node.node, False)
-        self.cache[node] = FunctionCall(
+    def _generate_artificial_rule_call(
+        self,
+        node: Any,
+        prefix: str,
+        rule_generation_func: Callable[[], str],
+        return_type: Optional[str] = None,
+    ) -> FunctionCall:
+        node_str = f"{node}"
+        key = f"{prefix}_{node_str}"
+        if key in self.cache:
+            name = self.cache[key]
+        else:
+            name = rule_generation_func()
+            self.cache[key] = name
```

Comment:
We can use `setdefault` for this.

```suggestion
	name = self.cache.setdefault(key, rule_generation_func())
```

Reply:
Unfortunately, this change is not correct: rule_generation_func will be called even in the case when key is already present in dict.
But rule_generation_func is not idempotent function.

On the same reason we can't write something like this:
name = cache.get(key, rule_generation_func())
cache[key] = name

Reply:
The difference here is that the rule generator function is always called even if the element is in the cache

Reply:
Hmm, you're both right, of course. I missed that.